tests/frontend/test_engine.py

```diff
@@ -166,209 +174,297 @@ def inspect():
+    def test_get_job_status(self, connection, monkeypatch):
+        """Tests a successful job status request."""
+        id_, status = "123", JobStatus.COMPLETE
 
-        some_params = {"param": MagicMock()}
-        result = starship_engine._run_program(program, args={}, compile_options={}, **some_params)
+        monkeypatch.setattr(
+            Connection,
+            "_get",
+            mock_return(mock_response(200, {"id": id_, "status": status.value})),
+        )
 
-        mock__get_blackbird.assert_called_once_with(program=program, param=some_params["param"],
-                args={}, compile_options={})
+        assert connection.get_job_status(id_) == status
 
-        assert result == mock_job.result.result.value
+    def test_get_job_status_error(self, connection, monkeypatch):
+        """Tests a failed job status request."""
+        monkeypatch.setattr(Connection, "_get", mock_return(mock_response(404, {})))
 
-    def test__run_program_fails(self, starship_engine, monkeypatch):
-        """
-        Tests that an Exception is raised when a job has failed.
-        """
-        mock_to_blackbird = MagicMock()
-        mock__get_blackbird = MagicMock()
-        program = MagicMock()
-        mock_job = MagicMock()
-        mock_job.is_complete = False
-        mock_job.is_failed = True
+        with pytest.raises(RequestFailedError):
+            connection.get_job_status("123")
 
-        monkeypatch.setattr("strawberryfields.engine.to_blackbird", mock_to_blackbird)
-        monkeypatch.setattr(starship_engine, "_get_blackbird", mock__get_blackbird)
-        monkeypatch.setattr(starship_engine, "_queue_job", lambda job_content: mock_job)
+    def test_get_job_result(self, connection, monkeypatch):
+        """Tests a successful job result request."""
+        result_samples = np.array([[1, 2], [3, 4]], dtype=np.int8)
 
-        some_params = {"param": MagicMock()}
+        with io.BytesIO() as buf:
+            np.save(buf, result_samples)
+            buf.seek(0)
+            monkeypatch.setattr(
+                Connection,
+                "_get",
+                mock_return(mock_response(200, binary_body=buf.getvalue())),
+            )
 
-        with pytest.raises(JobExecutionError) as e:
-            starship_engine._run_program(program, **some_params)
-        assert e.value.args[0] == str(mock_job.manager.http_response_data['meta'])
+        result = connection.get_job_result("123")
 
-    def test__run(self, starship_engine, monkeypatch):
-        """
-        Tests StarshipEngine._run, with the assumption that the backend is a hardware backend
-        that supports running only a single program. This test ensures that a program is compiled
-        for the hardware backend, is locked, is added to self.run_progs, that it is run and that
-        a Result object is returned populated with the result samples.
-        """
+        assert np.array_equal(result.samples.T, result_samples)
```

Comment:
This probably deserves some explanation. As the `Result` class is currently implemented, `Result.samples` has dimensions `(modes, shots)`; on the other hand, job results received from the remote backend have the shape `(shots, modes)`, so we transpose the samples here. Modifying `Result.samples` to return `(shots, modes)` appears to break various tests elsewhere in SF, so I opted to leave it unchanged. I also believe that the issue of sample dimensions is being (or has been) addressed on `master`, so it may be better to revisit this in another PR down the line.

Reply:
Thanks for the explanation! One thing to note is that there is an upcoming PR to be merged on this: #300

Reply:
Just mentioning that #300 is already merged (earlier today). It unified the sample returns for all simulator engines so that they return NumPy arrays with dimension `(shots, modes)`.

Reply:
Cool, thanks for the heads-up guys. I'm assuming we will have another substantial PR at some point to line up `demo` with `master`, so that may be a good time to get all the dimensions agreeing?